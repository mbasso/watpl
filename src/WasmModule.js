export default class WasmModule {
  constructor(source) {
    source.validate();
    this.source = source;
    this.string = this.source.toText({});
    this.buffer = this.source.toBinary({}).buffer;
    this.module = WebAssembly.compile(this.buffer);
  }

  instantiate = async (importObject = {}) =>
    WebAssembly.instantiate(await this.module, importObject);

  destroy = () => {
    this.source.destroy();
  };

  toString = () => this.string;
}
