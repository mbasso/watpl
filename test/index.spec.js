import watpl from '../src';
import WasmModule from '../src/WasmModule';

const source = `
  (module
    (func (param $lhs i32) (param $rhs i32) (result i32)
      get_local $lhs
      get_local $rhs
      i32.add)
    (export "add" (func 0))
  )
`;

describe('watpl', () => {
  it('should export wasm', () => {
    expect(watpl).toBeTruthy();
    expect(typeof watpl).toEqual('function');
  });

  it('should throw if an invalid module is provided', () => {
    expect(() =>
      watpl`
          (module
            (fun (param $lhs i32) (param $rhs i32) (result i32)
              get_local $lhs
              get_local $rhs
              i32.add)
            (export "add" (func 0))
          )
        `()
    ).toThrow();
  });

  it('should return module from string', () => {
    const module = watpl`${source}`();

    expect(module).toBeInstanceOf(WasmModule);
    expect(module.toString()).toMatchSnapshot();
  });

  it('should return a WasmModule', () => {
    const module = watpl`${source}`();

    expect(module).toBeInstanceOf(WasmModule);
  });

  it('should have an instantiate method', async () => {
    const spy = jest.spyOn(WebAssembly, 'instantiate');
    const importObject = { foo: () => null };
    const module = watpl`${source}`();
    const instance1 = await module.instantiate();
    const instance2 = await module.instantiate(importObject);

    expect(spy).toHaveBeenCalledTimes(2);
    expect(spy).toHaveBeenNthCalledWith(1, await module.module, {});
    expect(spy).toHaveBeenNthCalledWith(2, await module.module, importObject);

    expect(instance1).toMatchSnapshot();
    expect(instance2).toMatchSnapshot();
    expect(typeof instance1.exports.add).toEqual('function');
    expect(typeof instance2.exports.add).toEqual('function');
    expect(instance1).toBeInstanceOf(WebAssembly.Instance);
    expect(instance2).toBeInstanceOf(WebAssembly.Instance);
    spy.mockRestore();
  });

  it('should have a destroy method', () => {
    const module = watpl`${source}`();
    expect(() => {
      module.destroy();
    }).not.toThrow();
  });

  it('should have a toString method', () => {
    const module = watpl`${source}`();
    const stringSource = module.toString();

    expect(typeof stringSource).toEqual('string');
    expect(stringSource).toMatchSnapshot();
  });

  it('should have a string property', () => {
    const module = watpl`${source}`();
    const { string } = module;

    expect(typeof string).toEqual('string');
    expect(string).toMatchSnapshot();
  });

  it('should have a buffer property', () => {
    const module = watpl`${source}`();
    const { buffer } = module;

    expect(buffer).toMatchSnapshot();
  });

  it('should have a module property', async () => {
    const module = watpl`${source}`();
    const wasmModule = await module.module;

    expect(wasmModule).toBeInstanceOf(WebAssembly.Module);
    expect(wasmModule).toMatchSnapshot();
    expect(WebAssembly.Module.exports(wasmModule)).toMatchSnapshot();
    expect(WebAssembly.Module.imports(wasmModule)).toMatchSnapshot();
  });

  it('should compile module only once', async () => {
    const spy = jest.spyOn(WebAssembly, 'compile');
    const module = watpl`${source}`();
    await module.module;
    await module.module;

    expect(spy).toHaveBeenCalledWith(module.buffer);
    expect(spy).toHaveBeenCalledTimes(1);
    spy.mockRestore();
  });

  it('should interpolate strings', () => {
    const paramType = 'i32';
    const module = watpl`
      (module
        (func (param $lhs ${paramType}) (param $rhs ${paramType}) (result ${paramType})
          get_local $lhs
          get_local $rhs
          ${paramType}.add)
        (export "add" (func 0))
      )
    `();

    expect(module).toBeInstanceOf(WasmModule);
    expect(module.toString()).toMatchSnapshot();
  });

  it('should interpolate numbers', () => {
    const funcNumber = '0';
    const module = watpl`
      (module
        (func (param $lhs i32) (param $rhs i32) (result i32)
          get_local $lhs
          get_local $rhs
          i32.add)
        (export "add" (func ${funcNumber}))
      )
    `();

    expect(module).toBeInstanceOf(WasmModule);
    expect(module.toString()).toMatchSnapshot();
  });

  it('should interpolate functions', () => {
    const module = watpl`
      (module
        (func (param $lhs i32) (param $rhs i32) (result i32)
          get_local $lhs
          get_local $rhs
          i32.add)
        (export "${options => options.exportName}" (func 0))
      )
    `({
      exportName: 'add'
    });

    expect(module).toBeInstanceOf(WasmModule);
    expect(module.toString()).toMatchSnapshot();
  });
});
