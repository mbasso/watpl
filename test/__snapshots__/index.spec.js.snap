// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`watpl should have a buffer property 1`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  7,
  1,
  96,
  2,
  127,
  127,
  1,
  127,
  3,
  2,
  1,
  0,
  7,
  7,
  1,
  3,
  97,
  100,
  100,
  0,
  0,
  10,
  9,
  1,
  7,
  0,
  32,
  0,
  32,
  1,
  106,
  11,
]
`;

exports[`watpl should have a module property 1`] = `Module {}`;

exports[`watpl should have a module property 2`] = `
Array [
  Object {
    "kind": "function",
    "name": "add",
  },
]
`;

exports[`watpl should have a module property 3`] = `Array []`;

exports[`watpl should have a string property 1`] = `
"(module
  (func (;0;) (param $lhs i32) (param $rhs i32) (result i32)
    local.get $lhs
    local.get $rhs
    i32.add)
  (export \\"add\\" (func 0))
  (type (;0;) (func (param i32 i32) (result i32))))
"
`;

exports[`watpl should have a toString method 1`] = `
"(module
  (func (;0;) (param $lhs i32) (param $rhs i32) (result i32)
    local.get $lhs
    local.get $rhs
    i32.add)
  (export \\"add\\" (func 0))
  (type (;0;) (func (param i32 i32) (result i32))))
"
`;

exports[`watpl should have an instantiate method 1`] = `Instance {}`;

exports[`watpl should have an instantiate method 2`] = `Instance {}`;

exports[`watpl should interpolate functions 1`] = `
"(module
  (func (;0;) (param $lhs i32) (param $rhs i32) (result i32)
    local.get $lhs
    local.get $rhs
    i32.add)
  (export \\"add\\" (func 0))
  (type (;0;) (func (param i32 i32) (result i32))))
"
`;

exports[`watpl should interpolate numbers 1`] = `
"(module
  (func (;0;) (param $lhs i32) (param $rhs i32) (result i32)
    local.get $lhs
    local.get $rhs
    i32.add)
  (export \\"add\\" (func 0))
  (type (;0;) (func (param i32 i32) (result i32))))
"
`;

exports[`watpl should interpolate strings 1`] = `
"(module
  (func (;0;) (param $lhs i32) (param $rhs i32) (result i32)
    local.get $lhs
    local.get $rhs
    i32.add)
  (export \\"add\\" (func 0))
  (type (;0;) (func (param i32 i32) (result i32))))
"
`;

exports[`watpl should return module from string 1`] = `
"(module
  (func (;0;) (param $lhs i32) (param $rhs i32) (result i32)
    local.get $lhs
    local.get $rhs
    i32.add)
  (export \\"add\\" (func 0))
  (type (;0;) (func (param i32 i32) (result i32))))
"
`;
